File: Course_1-Algorithmic_Toolbox/Week-4/Excercise_Challenges/5_organizing_a_lottery/points_and_segments.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define PB push_back
#define MP make_pair
#define F first
#define S second
using namespace std;

bool cmp(pair<int, int> A, pair<int, int> B)
{
    return A.F < B.F;
}

vector<int> fast_count_segments(vector<pair<int, int> > v, vector<int> points)
{
    vector<int> cnt(points.size());
    for(int i = 0; i < points.size(); ++i){
        int con = 0;
        for(int j = 0; j < v.size(); ++j){
            if(v[j].F > points[i] || v[j].S < points[i])
                continue;
            else
                con++;
        }
        cnt[i] = con;
    }
    return cnt;
}

vector<int> naive_count_segments(vector<int> starts, vector<int> ends, vector<int> points)
{
    vector<int> cnt(points.size());
    for (size_t i = 0; i < points.size(); i++) {
        for (size_t j = 0; j < starts.size(); j++) {
            cnt[i] += starts[j] <= points[i] && points[i] <= ends[j];
        }
    }
    return cnt;
}

int main(void)
{
    int n, m;
    cin >> n >> m;

    vector<pair<int, int> > v;
    for(int i = 0, a, b; i < n; i++){
        cin >> a >> b;
        v.PB(MP(a, b));
    }
    sort(v.begin(), v.end(), cmp);

    vector<int> points(m);
    for(size_t i = 0; i < m; i++){
        cin >> points[i];
    }

    vector<int> cnt = fast_count_segments(v, points);
    for(size_t i = 0; i < cnt.size(); i++){
        cout << cnt[i] << ' ';
    }

    return 0;
}

File: Course_1-Algorithmic_Toolbox/Week-4/Excercise_Challenges/5_organizing_a_lottery/points_and_segments.py
from itertools import chain

def fast_count_segments(starts, ends, points):
    cnt = [0] * len(points)

    start_points = zip(starts, ['l'] * len(starts), range(len(starts)))
    end_points = zip(ends, ['r'] * len(ends), range(len(ends)))
    point_points = zip(points, ['p'] * len(points), range(len(points)))

    total_list = chain(start_points, end_points, point_points)
    total_list = sorted(total_list, key=lambda a: (a[0], a[1]))
    cont = 0

    for value, letter, index in total_list:
        if letter == 'l':
            cont += 1
        elif letter == 'r':
            cont -= 1
        else:
            cnt[index] = cont

    return cnt

def naive_count_segments(starts, ends, points):
    cnt = [0] * len(points)
    for i in range(len(points)):
        for j in range(len(starts)):
            if starts[j] <= points[i] <= ends[j]:
                cnt[i] += 1
    return cnt

if __name__ == '__main__':
    n, m = map(int, input().split())
    starts = []
    ends = []
    for _ in range(n):
        a, b = map(int, input().split())
        starts.append(a)
        ends.append(b)

    points = list(map(int, input().split()))
    cnt = fast_count_segments(starts, ends, points)
    for x in cnt:
        print(x, end=' ')
